cmd/evict/evict.go

```diff
@@ -0,0 +1,193 @@
+
+	GetNodeByName(ctx context.Context, name string) (*node.Resource, error)
+	GetVolume(ctx context.Context, namespaceID id.Namespace, uid id.Volume) (*volume.Resource, error)
+	GetNode(ctx context.Context, uid id.Node) (*node.Resource, error)
+	GetNamespaceByName(ctx context.Context, name string) (*namespace.Resource, error)
+	GetVolumeByName(ctx context.Context, namespaceID id.Namespace, name string) (*volume.Resource, error)
+
+	EvictReplica(ctx context.Context, namespaceID string, id string, deploymentID string) error
+	SetCordoned(ctx context.Context, nodeID id.Node, params *apiclient.SetCordonedRequestParams) error
+}
+
+type evictCommand struct {
+	config ConfigProvider
+	client Client
+
+	cordonTargetNode bool
+}
+
+// NewCommand configures the set of commands which are grouped by the "get" verb.
+func NewCommand(client Client, config ConfigProvider) *cobra.Command {
+	c := &evictCommand{
+		config: config,
+		client: client,
+	}
+
+	command := &cobra.Command{
+		Use:   "evict",
+		Short: "Evict removes a volume's replica from a node",
+		Example: `
+		$ storageos evict my-namespace my-volume my-node
+		$ storageos evict my-namespace my-volume my-node --cordon=false`,
+		Args: argwrappers.WrapInvalidArgsError(func(_ *cobra.Command, args []string) error {
+			if len(args) < 3 {
+				return errors.New("received too few args")
+			}
+			if len(args) > 3 {
+				return fmt.Errorf("received more args than expected: %v", args)
+			}
+			return nil
+		}),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			run := runwrappers.Chain(
+				runwrappers.RunWithTimeout(config),
+				runwrappers.EnsureNamespaceSetWhenUseIDs(config),
+				runwrappers.AuthenticateClient(config, client),
+			)(c.runWithCtx)
+
+			return run(context.Background(), cmd, args)
+		},
+		SilenceUsage: true,
+	}
+	command.Flags().BoolVar(&c.cordonTargetNode, "cordon", true, "cordon the target node before eviction")
+
+	return command
+}
+
+func (c *evictCommand) runWithCtx(ctx context.Context, cmd *cobra.Command, args []string) error {
+	useIDs, err := c.config.UseIDs()
+	if err != nil {
+		return err
+	}
+
+	nsID := id.Namespace(args[0])
+	vID := id.Volume(args[1])
+	targetNode := id.Node(args[2])
+
+	if !useIDs {
+
+		nn, err := c.client.GetNamespaceByName(ctx, string(nsID))
+		if err != nil {
+			return err
+		}
+		nsID = nn.ID
+
+		n, err := c.client.GetNodeByName(ctx, string(targetNode))
+		if err != nil {
+			return err
+		}
+		targetNode = n.ID
+
+		v, err := c.client.GetVolumeByName(ctx, nsID, string(vID))
+		if err != nil {
+			return err
+		}
+		vID = v.ID
+
+	}
+
+	vol, err := c.client.GetVolume(ctx, nsID, vID)
+	if err != nil {
+		return err
+	}
+
+	// Do some basic sanity checking to ensure we're not about to destroy a volume
+	if len(vol.Replicas) < 1 {
+		return fmt.Errorf("cannot evict volume if it has no replicas - data would be lost")
+	}
+
+	if vol.Master.Health != health.MasterOnline {
```

Comment:
Out of curiosity, wouldn't this be something you expect to be done server side?

Reply:
Yeah the control plane will just barf if we attempt to do this with an unhealthy master. That being said it seems sensible to also check here, as it means we can print a nice error